### PR TITLE
UNR-1000 make sure all dirty blueprints are compiled before schema generation

### PR DESCRIPTION
**Pull request by ieva-improbable**
#### Description

Make sure that all "dirty" blueprints are compiled before schema generation. If there are errors compiling blueprints, fail schema generation and output error messages specifying which blueprints were causing issues.

*NOTE:* this PR requires [these changes to the engine](https://github.com/improbableio/UnrealEngine/pull/81) to be merged in first!

#### Release note

"Bugfix:" Automatically compile all dirty blueprints before schema generation.

#### Tests

Prior to this PR, changes that were made to blueprints did not get picked up by schema generator unless the blueprints were explicitly compiled before generating schema. You can see that by making changes to a BP, generating schema, going back to the BP window and observing that the compile button has a question mark on it, showing that the BP is still dirty and has not been compiled. When you run the game, you can see that changes to the BP have not been picked up.

With these code changes, you'll see that after schema generation compile button will have a green tick on it for all the changed BPs, showing that they have been compiled. If any BPs have compilation errors, schema generation will fail and output error messages. If you now run the game, you can see that changes to BPs have been picked up.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 1]
```diff
@@ -331,6 +332,21 @@ void PreProcessSchemaMap()
 
 bool SpatialGDKGenerateSchema()
 {
+	// compile all changed blueprints
+	TArray<UBlueprint*> ErroredBlueprints;
+	bool PromptForCompilation = false;
+	FInternalPlayLevelUtils::ResolveDirtyBlueprints(PromptForCompilation, ErroredBlueprints);
+
+	if (ErroredBlueprints.Num() != 0)
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]:
```suggestion
	if (ErroredBlueprints.Num() > 0)
```

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 2]
```diff
@@ -331,6 +332,21 @@ void PreProcessSchemaMap()
...
+	{
+		UE_LOG(LogSpatialGDKSchemaGenerator, Error, TEXT("There were errors when compiling blueprints. Schema has not been generated."));
+		for (auto& Blueprint : ErroredBlueprints)
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]:
```suggestion
		for (const auto& Blueprint : ErroredBlueprints)
```

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp [hunk 3]
```diff
@@ -36,11 +36,9 @@ void FSpatialGDKEditor::GenerateSchema(FSimpleDelegate SuccessCallback, FSimpleD
 
 	PreProcessSchemaMap();
 
-	SchemaGeneratorResult = Async<bool>(EAsyncExecution::Thread, SpatialGDKGenerateSchema,
-		[this, bCachedSpatialNetworking, SuccessCallback, FailureCallback]()
-	{
-		if (!SchemaGeneratorResult.IsReady() || SchemaGeneratorResult.Get() != true)
-		{
+	AsyncTask(ENamedThreads::GameThread, [this, bCachedSpatialNetworking, SuccessCallback, FailureCallback]() {
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: Does it have to be on the GameThread?

Comment (ieva-improbable) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: Yes, `FKismetEditorUtilities::CompileBlueprint` is expected to run on the game thread as it does some copying/moving around of the game objects during compilation.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 4]
```diff
@@ -331,6 +332,21 @@ void PreProcessSchemaMap()
 
 bool SpatialGDKGenerateSchema()
 {
+	// compile all changed blueprints
+	TArray<UBlueprint*> ErroredBlueprints;
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Looking at this bit of code, could this all be moved into its own function and then called here?

Comment (ieva-improbable) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Pulled it out, lmk if it looks better now.

Comment (ieva-improbable) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Ignore me, talked to Mike and we agreed to take compilation out of the async task completely.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp [hunk 5]
```diff
@@ -36,6 +36,23 @@ void FSpatialGDKEditor::GenerateSchema(FSimpleDelegate SuccessCallback, FSimpleD
...
 
+	// Compile all dirty blueprints
+	TArray<UBlueprint*> ErroredBlueprints;
+	bool PromptForCompilation = false;
+	UEditorEngine::ResolveDirtyBlueprints(PromptForCompilation, ErroredBlueprints);
+
+	if (ErroredBlueprints.Num() > 0)
+	{
+		UE_LOG(LogSpatialGDKEditor, Error, TEXT("There were errors when compiling blueprints. Schema has not been generated."));
+		if (FailureCallback.IsBound())
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: Driveby: Could you switch all the IsBound()/Execute() logic in this file to ExecuteIfBound()?

**Comment by ieva-improbable** [comment]
The [required engine PR](https://github.com/improbableio/UnrealEngine/pull/81) has now been merged into `4.20-SpatialOSUnrealGDK` branch. @m-samiec let me know if you're ok with this PR going in now.

**Comment by m-samiec** [comment]
As this change requires an engine rebuild - lets hold off merging until the BK pipeline has been merged. Will update here when it's gone in.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp [hunk 6]
```diff
@@ -36,22 +36,30 @@ void FSpatialGDKEditor::GenerateSchema(FSimpleDelegate SuccessCallback, FSimpleD
 
 	PreProcessSchemaMap();
 
+	// Compile all dirty blueprints
+	TArray<UBlueprint*> ErroredBlueprints;
+	bool PromptForCompilation = false;
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]:
```suggestion
	bool bPromptForCompilation = false;
```

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp [hunk 7]
```diff
@@ -36,22 +36,30 @@ void FSpatialGDKEditor::GenerateSchema(FSimpleDelegate SuccessCallback, FSimpleD
...
+	UEditorEngine::ResolveDirtyBlueprints(PromptForCompilation, ErroredBlueprints);
+
+	if (ErroredBlueprints.Num() > 0)
+	{
+		UE_LOG(LogSpatialGDKEditor, Error, TEXT("There were errors when compiling blueprints. Schema has not been generated."));
+		FailureCallback.ExecuteIfBound();
+		GeneralProjectSettings->bSpatialNetworking = bCachedSpatialNetworking;
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: Curious why this line is needed?

Comment (ieva-improbable) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: Not 100% sure, but it would have always been called without my changes (see [here](https://github.com/spatialos/UnrealGDK/blob/14b3a98f5b13f0fbbc9c97a0f0047dcdc778bd25/SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp#L64)), so I didn't want to accidentally break anything. Let me know if you think it is safe to take it out.
